大規模改修

gomagom opened the pull request:
- メインメニューを作り直し
- ゲーム本体を作り直し
- 曲選択画面やリザルト画面なども追加

Commit:
作り直し中

Commit:
修正中

Commit:
CSVファイル読み込みとノーツの設置、画面の基本要素を作成

CSVファイルインポートを実装(仮)
譜面データをもとにノーツインスタンスを生成し、流れてくるように設定
レーンや判定ラインを設置

Commit:
初期オーバーレイ画面と判定、画面のアニメーション類を追加。
その他多数の調整、修正、要素追加を含む

Commit:
色々追加

Commit:
ゲーム画面のデザインを更新
コンボ表記を実装

Commit:
リザルト画面の追加とデザインの調整、ファイルの整理など

リザルト画面を実装しました
ゲーム開始時の画面のデザインを調整中です
いらないファイルや記述を消しました
効果音を実装しました

Commit:
デザインを変更

Commit:
音量調節機能を追加し、見た目を調整

Commit:
少し修正

Commit:
キーコンフィグを追加

Commit:
タイトル画面において、音量やキーコンフィグの設定を実装

yamakenji24 commented on js/game/game.js:
```diff
@@ -1,260 +1,159 @@
-window.onload = function() {
-    setLane();
-	getCSV();
+window.onload = () => {
```
[MUST] window.onloadのようなDomイベントの操作は、EventListenerで制御する方が現状推奨されています。
例えば、window.onloadの場合は、`window.addEventListener('load', () => {})` 。

gomagom replied:
一通りEventListener制御に変更しました。

yamakenji24 commented on js/index/script.js:
```diff
@@ -0,0 +1,76 @@
+    bgmVolume: document.getElementById('BGM-slider'),
+    seVolume: document.getElementById('SE-slider')
+}
+const KEY_CONFIG = document.getElementsByClassName('key-config');
```
[IMO] 大文字宣言は基本的に、中の値が変わらないことを保証するためにつけるものです。
jsのオブジェクトの場合、変更できるのでどうなんだろうというのがお気持ちです。

gomagom replied:
配列も下のような場合小文字にしたほうがいいですか？
const BACK_LANE = [];

yamakenji24 commented on js/index/script.js:
```diff
@@ -0,0 +1,76 @@
+    }
+};
+
+window.onbeforeunload = () => {
```
[IMO] 同様に、eventlistener でイベントを制御した方が適切だと思います！

yamakenji24 replied:
[Q] リロード等をしても、再プレイできるようにするための実装ですか？

gomagom replied:
そうです！BGMやSEのボリューム、キーコンフィグなどをローカルストレージに保存し、再びアクセスされた際に読み込むようにしてます。

yamakenji24 commented on js/index/script.js:
```diff
@@ -0,0 +1,76 @@
+SETTING.addEventListener('click', () => OVERLAY.style.display = "block");
+RETURN.addEventListener('click', () => OVERLAY.style.display = "none");
+
+window.onload = () => {
```
[MUST] window.onloadのようなDomイベントの操作は、EventListenerで制御する方が現状推奨されています。
例えば、window.onloadの場合は、window.addEventListener('load', () => {}) 。

gomagom replied:
修正しました

yamakenji24 commented on js/index/script.js:
```diff
@@ -0,0 +1,76 @@
+    }
+    if (storage.seVolume >= 0) {
+        SOUND.seVolume.value = storage.seVolume * 50;
+    }
```
[IMO] 例外発生後に、例外発生しない前提での処理が記述されているように思うので、try catchの中で書いた方がいいように思いました。

gomagom replied:
if文をtryの中に入れました！

yamakenji24 commented on js/game/game.js:
```diff
@@ -1,260 +1,159 @@
+        } else {
+            SOUND[key] = storage[key];
+        }
+    }
```
[IMO] 例外発生後に、例外発生しない前提での処理が記述されているように思うので、try catchの中で書いた方がいいように思いました。

gomagom replied:
書き換えました。

yamakenji24 replied:
さらに付き加えると、例外発生後のエラーハンドリングがあるとなお良いかなーと思いました。
localstorageからほぼ取ってこれるように設計されているとは思いますが、取ってこれなかった時ようにトップ画面に戻る/曲を選択する...等々のハンドリングがあっても良いのかなと(なくても動きますが)

yamakenji24 commented on js/game/function.js:
```diff
@@ -0,0 +1,296 @@
+function convertCSVtoArray(str) {
+    const RESULT = [];
+    const TMP = str.split('\r\n');                                  // 改行位置で分割して格納
+    TMP.forEach((val, index) => RESULT[index] = val.split(','));    // ','の位置で分割して格納
```
[IMO] array.mapを使った方がコードがスッキリすると思います!
もし、可能そうであれば、splitと同時に数値型への変換も検討してみてください。
`val.split.map(Number)` みたいなのはできないかな・・？

gomagom replied:
array.mapで修正しました。
val.split(',').map(Number)で数値型への変換も同時にできました！

yamakenji24 replied:
LGTM!

yamakenji24 commented on js/game/class.js:
```diff
@@ -0,0 +1,207 @@
+class JudgeLine {
```
[Q] JudgeLineが何を受け持つクラスなのか、よくわかりませんでした。

gomagom replied:
canvas内の判定用ラインの座標情報などを保持し、自身を描画するって感じですかね。ノーツの座標もこれを基準に求めています。

yamakenji24 commented on js/game/game.js:
```diff
@@ -1,260 +1,159 @@
+    great: 0,
+    good: 0,
+    bad: 0
+};
```
[IMO] Scoreを管理するクラスを作ってみるのもありかなと思いました。

gomagom replied:
スコアを管理するGameScoreManagerを作ってみました

yamakenji24 commented on js/game/function.js:
```diff
@@ -0,0 +1,296 @@
+    gameActive = false;
+
+    SCORE.point = Math.floor(SCORE.point);
+    const INFO = document.getElementsByClassName('result-info');
```
[IMO] ここのInfoの変数名が、game.jsで定義されているものと同等になるので、例えばResultInfoのように、違いがわかるといいなとおもいました。

gomagom replied:
変数名をINFO→RESULT_INFOとしました。

yamakenji24 commented on js/game/function.js:
```diff
@@ -0,0 +1,296 @@
+        }
+    });
+
+    window.onbeforeunload = () => {
```
[IMO] 下記同様

gomagom replied:
修正しました。

yamakenji24 commented on js/game/game.js:
```diff
@@ -1,260 +1,159 @@
+window.onload = () => {
+    ELEMENT.uploadMusicFile.onchange = () => {
+        ELEMENT.uploadStr[1].textContent = ELEMENT.uploadMusicFile.files[0].name;
+        console.log(ELEMENT.uploadMusicFile.files[0].name);
```
[NITS] `console.log` 残ったままになってる 👀

gomagom replied:
ありゃ、すいません。消しました。

Amakuchisan commented on js/game/class.js:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    drawGrade(grade) {
+        const STRING = ['Parfect!', 'Great!', 'Good!', 'Bad...'];
```
```suggestion
        const STRING = ['Perfect!', 'Great!', 'Good!', 'Bad...'];
```
パーフェクト……

yamakenji24 replied:
パーフェクト!

gomagom replied:
あっ...修正しました

Amakuchisan reviewed (changes requested):
Perfect(改修後のUIすごくなってました……)

yamakenji24 reviewed (changes requested):
全体的な感想ですが、ゲームとしてはよくできていると思います。
前回のものから非常にバージョンアップされており、わかりやすくなっていたと思います。
ファイルのわけ方などが少し雑かなーっていう雰囲気を感じました。
１クラス１ファイル等、何の実装を行っているファイルか、わかりやすい方が望ましいです。 

[MUST]で指摘したポイントだけは、修正をお願いします。
他の部分に関してですが、現状の推奨事項やこうした方が良さそうかも？ 的なものも含まれていますので、検討をお願いします。

Commit:
ファイル分けを見直しました

Commit:
・Domイベント操作をEventListenerで制御するよう変更
・例外処理部分を少し修正
・その他細かい修正

Commit:
csvファイルの読み込みに関して、array.mapを使用することでコードを簡略化

Commit:
- オブジェクト名を大文字から小文字に変更しました
- ゲームスコアを管理するクラスを作りました

Commit:
少し修正

Commit:
・ローカルストレージからデータを取ってくる際、エラーが起こった場合はページ再読み込みを実施
・FireFoxでも見れるレベルにCSSを修正
・譜面や楽曲データの読み込み終了までゲームの実行を待つように変更
・その他余分な記述などを修正

yamakenji24 commented on js/game/init.js:
```diff
@@ -0,0 +1,159 @@
+
+// 譜面データをレーン毎に切り分ける
+function cutMusicScore(arr) {
+    const DATA = new Array(4);
```
[IMO] DATAという変数名が曖昧かなと思います。具体的な格納されているデータを表すような命名が良いかなと思います。(scoreData, musicDataなどなど)

gomagom replied:
譜面の管理もクラスにまとめました。

yamakenji24 commented on js/game/loop.js:
```diff
@@ -0,0 +1,127 @@
+    });
+
+    window.addEventListener('blur', () => {
+        if (gameActive && !gameFinish) {
```
[Q] 全体的に、gameFinishがなくても、gameActiveだけでいけそうなのかなという雰囲気を感じるんですが、どうなんでしょうか？

gomagom replied:
ポーズ画面を出すときに、gameActiveだけだとリザルト画面に移行しているかどうか分からないので、今のやり方だと必要ですね。

yamakenji24 commented on js/game/main.js:
```diff
@@ -0,0 +1,136 @@
+    bgmVolume: 0.25,
+    seVolume: 1,
+    data: null
+};
```
[IMO] soundに関連する処理がいくつか見られるので、クラスでまとめちゃっても良いかもと思いました。
[Q] sound.bgm.volume は見た感じ使ってなさそう...?

gomagom replied:
sound関連はクラスでまとめました。sound.bgm.volumeはbgmの音量変更で使っています。loop.jsの79行目あたりですかね。

yamakenji24 replied:
把握しました 
sound.bgm は new Audioに紐づいているんですね

yamakenji24 commented on js/game/main.js:
```diff
@@ -0,0 +1,136 @@
+    bpm: 0,
+    musicL: 0,
+    musicStart: 0
+};
```
[Q] ここのinfo、代入だけされてて特に他で使ってそうな場所が見当たらないのですが、将来的に使う予定というニュアンスで？

gomagom replied:
デバックで使ったり、画面表示用に使うかなと思っていたのですが、使わなそうなので削除しました。

yamakenji24 reviewed (changes requested):
全体的に良くなってきていると思います!

Commit:
リスタート時にBGMがずれる問題を修正

Commit:
譜面を管理するクラスを作りました
サウンド関連をまとめて管理するクラスを作りました
ゲーム開始時にキーを押下してから処理が始まるようにしました
譜面とBGMがずれる問題を修正しました
Safariに対応しました
一部変数名などを変更しました
不要な記述を整理しました
その他いくつかの修正

Commit:
デバッグ表示を削除

Commit:
UIを改良しました
ある程度のウインドウ比率まではレイアウトが崩れないようにしました
ゲームの開始画面のUIを変更しました
設定項目にノーツのディレイとスピード比率を追加しました

Commit:
軽微な修正

Commit:
バグの修正他

Commit:
コードを整理
